### DISCUSSION TITLE
ObjectCreationExceptionWithPath for recursive type

### DISCUSSION THREAD

**Discussion by peter-bannerflow** [AutoFixture/AutoFixture#1245] [Q&A] [answered]
I have an issue with the following types:

```
public class Group 
{
   public readonly ICollection<Person> Members;
   
   public Group(ICollection<Person> members)
   {
      Members = members;
   }
}

public class Person
{
   public readonly ICollection<Person> Children;
  
   public Person(ICollection<Person> children) 
   {
      Children = children;
   }
}
```

When used in a test, I have configured my fixture to deal with the recusion in the Person-class by doing the following:
```
fixture.Behaviors.OfType<ThrowingRecursionBehavior>().ToList().ForEach(b => 
   fixture.Behaviors.Remove(b));
fixture.Behaviors.Add(new OmitOnRecursionBehavior());
```

However, in my test where I create a Group-object using the fixture, I get the following `ObjectCreationExceptionWithPath `-exception:
```
AutoFixture.ObjectCreationExceptionWithPath : AutoFixture was unable to create an instance from Person, most likely because it has no public constructor, is an abstract or non-public type.
```

**Comment by dammejed** [comment]
A similar question/issue was raised on StackOverflow: https://stackoverflow.com/questions/63377109/why-cant-autofixture-create-this-class

What's curious is that creating `Person` through Autofixture works just fine, but it seems to be unable to handle creation of the `Group` type, even though Autofixture should (presumably) be able to satisfy the `ICollection<Person>` parameter for that type's constructor.

**Comment by Hykke** [comment, marked as answer]
I believe the problem is that the RecursionGuard class will return an OmitSpecimen result when it reaches the recursion limit. This result is handled beautifully if it's a property assignment, but if it happens to be for a parameter for a consttructor, the OmitSpecimen class is not handled, and it will try to call the constructor with the OmitSpecimen object - this will of course fail

I think this is a case where NullRecursionBehavior can solve the problem - but of course maybe the logic around OmitSpecimen can be improved

**Comment by Z10yTap0k** [comment]
Why NullRecursionBehavior is not working when update to latest version ?

AutoFixture 3.51.0 AutoFixture.AutoMoq 3.51.0 - test **pass**
AutoFixture 4.15.0 AutoFixture.AutoMoq 4.15.0 - test **fail**
xunit used 2.4.0 version

Fail message:
AutoFixture.ObjectCreationExceptionWithPath
AutoFixture was unable to create an instance from TestProject1.UnitTest1+TestRecursion, most likely because it has no public constructor, is an abstract or non-public type.
Request path:
	TestProject1.UnitTest1+TestRecursion

Code:
```
    public class UnitTest1 : Fixture
    {
        public class TestRecursion
        {
            public readonly TestRecursion Prev;

            public TestRecursion(TestRecursion prev)
            {
                Prev = prev;
            }
        }
        
        public UnitTest1()
        {
            Customize(new AutoMoqCustomization());
        }
        
        [Fact]
        public void Test1()
        {
            Behaviors.OfType<ThrowingRecursionBehavior>()
                .ToList().ForEach(b => Behaviors.Remove(b));
            Behaviors.Add(new NullRecursionBehavior());
            
            var test = this.Create<TestRecursion>();
            Assert.NotNull(test);
            Assert.Null(test.Prev);
        }
    }
```

Thank for answer;

**Comment by aivascu** [comment]
@Z10yTap0k this behavior is intended by design. 
A constructor that contains a single parameter of the same type as the declaring type is considered by convention a "copy constructor". 
AutoFixture cannot satisfy this request since these type of constructors require a non-null value to copy its contents from.

Also please do not hijack other questions. You can open a new question in the Q&A discussions section.

**Reply by Z10yTap0k** [reply]
@aivascu , great, but this behavior even if the model has multiple parameters.
I migrate project from v2 to v4.
In version 2.x of Autofixture we simple add NullRecursionBehavior in Fixture ctor. 
But in v4 this trick is not working.
Implementation of the ISpecimenBuilder for TestRecursion also did not bring success.

**Reply by aivascu** [reply]
@Z10yTap0k the behavior you're experiencing is intended by design. This change was introduced in #937 to prevent AutoFixture from failing while resolving recursive constructor invocations. Basically any constructor that requires an instance of the same type as the constructed type is considered a copy constructor, event if it has other parameters.

Allowing the class to receive `null` in the copied argument is bad practice (likely even a bug) as the copy operation cannot be performed.

If you insist on having this design, you could implement a [custom constructor query](https://stackoverflow.com/a/12425942/2542567) that reverts the change introduced by #937 and then customize the fixture so that it makes use of the constructor query (see below), while applying the `NullRecursionBehavior` as you've done in the example above.

```cs
fixture.Customize<TestRecursion>(
    x => x.FromFactory(
        new MethodInvoker(
            new CustomConstructorQuery())));
```

Please create a new thread next time.
